csrc/runtime/executor.cpp

```diff
@@ -1267,7 +1267,53 @@ KernelArgumentHolder KernelExecutor::run(
               << ", occupancy=" << oss.str() << std::endl;
     }
 
-    if (!compiled_kernel_->kernel()->summary().has_cooperative_grid_reduction) {
+    const auto& kernel_summary = compiled_kernel_->kernel()->summary();
+
+    // cluster reduction uses DSMEM
+    if (kernel_summary.has_cluster_reduction) {
+      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchKernelEx");
+      CUlaunchConfig config = {};
+      config.gridDimX = launch_params_.gdimx();
+      config.gridDimY = launch_params_.gdimy();
+      config.gridDimZ = launch_params_.gdimz();
+      config.blockDimX = launch_params_.bdimx();
+      config.blockDimY = launch_params_.bdimy();
+      config.blockDimZ = launch_params_.bdimz();
+      config.sharedMemBytes = launch_params_.smem();
+      config.hStream = stream;
+
+      CUlaunchAttribute attribute;
+      attribute.id = CU_LAUNCH_ATTRIBUTE_CLUSTER_DIMENSION;
+      attribute.value.clusterDim.x = launch_params_.gdimx();
+      attribute.value.clusterDim.y = 1;
+      attribute.value.clusterDim.z = 1;
+      config.attrs = &attribute;
+      config.numAttrs = 1;
+      // TO request more than 8 clusters, need to set non-portable cluster size
```

Comment:
```suggestion
      // To request more than 8 clusters, need to set non-portable cluster size
```

Reply:
You are right. Should be 8 CTAs per cluster.